File: hw2/p4/p4.cpp
// NAME       : Aaron Graham
// COURSE     : NERS 544
// ASSIGNMENT : Homework 2
// PROBLEM    : Problem 4
// DATE       : April 1, 2015

#define PI 3.14159265358979323

#include<iostream>
#include<cstdlib>
#include<time.h>
#include<cmath>
#include<iomanip>

using namespace std;

double normRand(void)
{
  return static_cast<double>(rand())/static_cast<double>(RAND_MAX);
}

int main(void)
{

  const int N=1e6;
  double Ex[5] = {0.0, 0.0, 0.0, 0.0, 0.0};
  double tau, x;

  for (int n = 0; n < 5; n++)
  {
    // Even n
    if (n%2 == 0)
    {
      for (int i = 0; i < N; i++)
      {
        tau = sqrt(-log(normRand()))*cos(2.0*PI*normRand());
        x = 1.0;
        for (int k = 0; k < n/2; k++)
        {
          x *= normRand();
        }
        x = sqrt(-log(x) + tau*tau);
        Ex[n] += x;
      }
    }
    // Odd n
    else
    {
      for (int i = 0; i < N; i++)
      {
        x = 1.0;
        for (int k = 0; k < (n+1)/2; k++)
        {
          x *= normRand();
        }
        x = sqrt(-log(x));
        Ex[n] += x;
      }
    }
    Ex[n] /= N;
    cout << "For n=" << n << ", E[x]=" << setprecision(9) << Ex[n] << endl;
  }

  return 0;

}

File: hw2/p4/p4.m
close all, clear all, clc;
format long;

N = 1e6;
Ex = [0, 0, 0, 0, 0];
minn = 0;
maxn = 4;

for n = minn:maxn
  % Even n
  if (mod(n,2) == 0)
    for j = 1:N
      tau = sqrt(-log(rand()))*cos(2*pi*rand());
      x = 1;
      for k = 1:(n/2)
        x = x*rand();
      end
      x = sqrt(-log(x) + tau*tau);
      Ex(n+1) = Ex(n+1) + x;
    end
  % Odd n
  else
    for j = 1:N
      x = 1;
      for k = 1:((n+1)/2)
        x = x*rand();
      end
      x = sqrt(-log(x));
      Ex(n+1) = Ex(n+1) + x;
    end
  end
end

Ex = Ex/N
